core/src/main/scala/kafka/log/OffsetMap.scala

```diff
@@ -68,90 +120,164 @@ class SkimpyOffsetMap(val memory: Int, val hashAlgorithm: String = "MD5") extend
+  }
   
   /**
    * Associate this offset to the given key.
-   * @param key The key
-   * @param offset The offset
+   * @param record The record
+   * @return success flag
    */
-  override def put(key: ByteBuffer, offset: Long): Unit = {
+  override def put(record: Record): Boolean = {
     require(entries < slots, "Attempt to add a new entry to a full offset map.")
+
+    val key = record.key
+    val offset = record.offset
+    val currVersion = extractVersion(record)
+
+    lastOffset = offset
     lookups += 1
     hashInto(key, hash1)
+
     // probe until we find the first empty slot
     var attempt = 0
     var pos = positionOf(hash1, attempt)  
-    while(!isEmpty(pos)) {
+    while (!isEmpty(pos)) {
       bytes.position(pos)
       bytes.get(hash2)
-      if(Arrays.equals(hash1, hash2)) {
+      if (Arrays.equals(hash1, hash2)) {
+        // we found an existing entry, overwrite it and return (size does not change)
+        if (this.compactionStrategy != CompactionStrategy.OFFSET) {
+          // read previous value by skipping offset
+          bytes.position(bytes.position() + longByteSize)
+          val foundVersion = bytes.getLong()
+          if (foundVersion > currVersion) {
+            // map already holding latest record
+            return false
+          }
+
+          // reset the position to start of offset
+          bytes.position(bytes.position() - (2*longByteSize))
+        }
+
         // we found an existing entry, overwrite it and return (size does not change)
         bytes.putLong(offset)
-        lastOffset = offset
-        return
+        if (this.compactionStrategy != CompactionStrategy.OFFSET) {
+          bytes.putLong(currVersion)
+        }
+
+        return true
       }
+
       attempt += 1
       pos = positionOf(hash1, attempt)
     }
+
     // found an empty slot, update it--size grows by 1
     bytes.position(pos)
     bytes.put(hash1)
     bytes.putLong(offset)
-    lastOffset = offset
+    if (this.compactionStrategy != CompactionStrategy.OFFSET) {
+      bytes.putLong(currVersion)
+    }
+
     entries += 1
+    true
   }
-  
+
   /**
    * Check that there is no entry at the given position
    */
   private def isEmpty(position: Int): Boolean = 
-    bytes.getLong(position) == 0 && bytes.getLong(position + 8) == 0 && bytes.getLong(position + 16) == 0
+    bytes.getLong(position) == 0 && bytes.getLong(position + longByteSize) == 0 && bytes.getLong(position + (2*longByteSize)) == 0
 
   /**
-   * Get the offset associated with this key.
+   * Checks to see whether to retain the record or not
+   * @param record The record
+   * @return true to retain; false not to
+   */
+  override def shouldRetainRecord(record: Record): Boolean = {
+    if (this.compactionStrategy != CompactionStrategy.OFFSET) {
+      val foundVersion = getVersion(record.key)
+      val currentVersion = extractVersion(record)
+      // use version if available & different otherwise fallback to offset
```

Comment:
yes, that is handled in the put method... here we are checking whether to retain the record or not based on the already built map.

Reply:
The case I am thinking is the following. Suppose in log segment 1, we have (offset, key, timestamp) (10, 1, 100). We do a round of timestamp compaction and keep this record. In log segment 2, we have (20, 1, 99). When building the map, we only load the timestamp 99 for key 1. This means that during cleaning, offset 20 won't be removed, but it seems that it should be. Also, according to the KIP, we need to handle LEO specially. If the record for LEO is going to be removed, we need to (1) retain an empty batch to keep LEO if the message format is 2 or above; (2) keep the message if the format is below 2.

Reply:
Also there's one optimization we can do: after we've called `getVersion(record.key)`, we can actually avoid a second search if we have to call `getOffset(..)` below in 247 since we know the offset can be read together with the version: maybe we can let `getVersion()` to return the combo of "offset, version", for example.